### PR TITLE
Implement MGU

### PR DESCRIPTION
**Pull request by unnonouno**
I implemented Minimal Gated Unit.
http://arxiv.org/abs/1603.09420

### CODE REVIEW

**Comment by delta2323** [comment]
Do you think we should explicitly name `StatelessMGU` instead of `MGU` in the same as LSTM?

File: chainer/links/connection/mgu.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+    def to_gpu(self, device=None):
+        super(StatefulMGU, self).to_gpu(device)
+        if self.h is not None:
+            self.h.to_gpu(device)
+
+    def set_state(self, h):
+        assert isinstance(h, chainer.Variable)
+        h_ = h
+        if self.xp == numpy:
```
Comment (delta2323) [inline chainer/links/connection/mgu.py]: Use `is` to compare modules.

File: chainer/links/connection/mgu.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        else:
+            h_.to_gpu()
+        self.h = h_
+
+    def reset_state(self):
+        self.h = None
+
+    def __call__(self, x):
+        if self.h is None:
+            n_batch = x.data.shape[0]
```
Comment (delta2323) [inline chainer/links/connection/mgu.py]: Use `len()` to extract first element of a tuple.

File: chainer/links/connection/mgu.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            h_data = self.xp.zeros(
+                (n_batch, self.state_size), dtype=numpy.float32)
+            h = chainer.Variable(h_data)
+        else:
+            h = self.h
+
+        self.h = MGU.__call__(self, h, x)
```
Comment (delta2323) [inline chainer/links/connection/mgu.py]: I think it is better to extract `MGU.__call__` as a `Function` class like `MGUFunction` that takes `W_f`, `W_h`, `h`, and `x` and returns `h_new`.

File: tests/chainer_tests/links_tests/connection_tests/test_mgu.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            -1, 1, (3, self.in_size)).astype(numpy.float32)
+        self.gy = numpy.random.uniform(
+            -1, 1, (3, self.out_size)).astype(numpy.float32)
+
+        self.mgu = links.StatefulMGU(self.in_size, self.out_size)
+
+    def check_forward(self, x_data):
+        x = chainer.Variable(x_data)
+        self.mgu(x)
+        self.mgu(x)
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_mgu.py]: Why we call the method twice?

**Comment by unnonouno** [comment]
We need to make a policy about names of RNN algorithms first.

File: chainer/links/connection/mgu.py (hunk 3 above)
Comment (unnonouno) [inline chainer/links/connection/mgu.py]: It's not a `Function` but a python function. I mean I didn't implement backward propagation.
Is it really useful?

File: tests/chainer_tests/links_tests/connection_tests/test_mgu.py (hunk 4 above)
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_mgu.py]: OK, make sense

File: chainer/links/connection/mgu.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        )
+
+    def __call__(self, h, x):
+        f = sigmoid.sigmoid(self.W_f(concat.concat([h, x])))
+        h_bar = tanh.tanh(self.W_h(concat.concat([f * h, x])))
+        h_new = linear_interpolate.linear_interpolate(f, h_bar, h)
+        return h_new
+
+
+class StatefulMGU(MGU):
```
Comment (delta2323) [inline chainer/links/connection/mgu.py]: It have the same problem as one in SGD and DSGU in #1115. `StatefulMGU` may not be the subclass of `MGU`.

**Comment by delta2323** [comment]
@unnonouno How about changing the name from `SGU` to `StatelessSGU` so that we can create the alias `SGU` afterwards? We can merge this PR once we determine the names.

**Comment by delta2323** [comment]
As the naming rule of RNN modules is determined. We can merge this PR. Could you resolve the conflict?

**Comment by niboshi** [comment]
Resolved conflict and rewrote to use `use_scope`.

**Comment by delta2323** [comment]
jenkins, test this please.

**Comment by delta2323** [comment]
test passed

**Pull request merged**